### DISCUSSION TITLE
AO3 rate limit solution?

### DISCUSSION THREAD

**Discussion by IceRed-Drone** [nianeyna/ao3downloader#103] [Ideas]
Hi, I just started using the program. I intend to download large numbers of fics, so time is an important factor.
I know that AO3 rate limits mean that there have to be pauses in the download, but at the moment it's pausing 1-2 times per page for 5 minutes. I've previously used the chrome extension Simple Mass Downloader and can download at least 100 fics without running into a timeout request. With that extension I can set it to wait between each download, and I usually set it for 10 seconds; I'm not sure how this program currently does things, but would implementing a small waiting period like that make timeouts appear less frequently?

**Comment by verotheelf** [comment]
Hmm getting the code to stop is really easy with something like a sleep function, it's just a matter of figuring out the best place to add the pause. Would you want it to wait after each story is completely downloaded or also wait in-between downloading different file types (if you were downloading more than one)?  

If between works I think adding `sleep(10)` somewhere around [here](https://github.com/nianeyna/ao3downloader/blob/243e2d0462e438fdcab261c597593e0f3f709a35/ao3downloader/ao3.py#L204) should be fine.

If between file types then you can try it [here](https://github.com/nianeyna/ao3downloader/blob/243e2d0462e438fdcab261c597593e0f3f709a35/ao3downloader/ao3.py#L183)

Change the number in `sleep(10)` to your preferred amount of seconds. And add this to the top of the file `from time import sleep`. Please let us know if you find an optimal rate!

Of course this will only help with actions that cause timeouts specifically because of their call for downloads. Since something like updating requires a lot of interaction beyond just downloading, a different solution would be required.

**Reply by nianeyna** [reply]
I would actually put it just before [this line](https://github.com/nianeyna/ao3downloader/blob/243e2d0462e438fdcab261c597593e0f3f709a35/ao3downloader/repo.py#L73) - there's no advantage, I think, to only waiting after a fic has been fully downloaded, or anything like that - a hit to ao3 is a hit to ao3, if you're going to wait you should wait on each one. That said I *am* working on something to make this configurable, but I don't know when it'll be out.

**Reply by verotheelf** [reply]
Ah that's where I thought of putting it at first as it would work with the updating too but I second guessed whether I actually remembered how the code worked. Thank you!

**Reply by IceRed-Drone** [reply]
From what I understand, the rate limiter is about how frequently you're "clicking" on the site - the warning it gives even says you're clicking too fast - so it shouldn't really matter if the download is finished so much as how frequently requests are made to the site.

**Reply by IceRed-Drone** [reply]
I had to wait a few days before testing, but this solution did indeed work. At the moment I've got it set for 5 seconds, made a little custom message so I can see the progress of each fic, and it hasn't had any issues for 12 pages.

**Reply by Kyther** [reply]
Hmm, metadata grabs with a 5-second wait still triggered the rate limit sometime after page 60.  6 seconds lasted for 70 pages that time before requesting a 167-second break.  I'll have to see how it goes with fics.  It seems that they perhaps have a "total number of hits in a time period" metric somehow, and it's still hitting that eventually, just takes longer when I slow it down.  I'm trying 7 seconds now to see if it'll get through the rest of this grab without stopping (assuming it doesn't try to go on forever as well, lol).

EDIT:  It did not, and I switched to 8 seconds, but there were too few pages left in that particular grab to test if that avoided the wait time.  However, 8 seconds is doing great with the fics so far - well past the 20 file mark where I never managed to get anything past without triggering the wait.

**Reply by IceRed-Drone** [reply]
5 and 6 seconds do indeed eventually trigger the rate limiter (though it still runs for a few pages and only paused for under 60 seconds for me). After a few days of testing, 7 seconds seems to be ideal, I haven't had any pauses even when running all night.

**Reply by Kyther** [reply]
I'm guessing something involved in downloading files slows it down just that extra bit, then, because even 7 seconds triggered the wait time with grabbing metadata.  I'll try the metadata on 8 seconds next time I have to grab a big set - and I have the current run of fics downloading going on 7 secs, but it seems to actually result in somewhere between 7 and 8 seconds.  I'm guessing that's the sweet spot.  Metadata grabbing goes so quickly that I'm wondering if that 7 seconds only is just that hair too fast for it, because it doesn't have the time factored (only fractions of a second but still) in getting the database to respond with the files.
